### PR TITLE
Keep the stored attachment metadata when rebuilding thumbnails

### PR DESCRIPTION
**Pull request by toolshedlabs-hash**
Rebuilding thumbnails replaces the whole attachment metadata record, so anything the plugin does not write itself is lost. On any image WordPress scaled at upload time that includes `original_image`, and once that key is gone the full size file is stranded on disk with nothing pointing at it.

### What happens

The ajax handler does this:

```php
wp_update_attachment_metadata( $id, wp_generate_attachment_metadata_custom( $id, $fullsizepath, $thumbnails ) );
```

`wp_update_attachment_metadata()` overwrites `_wp_attachment_metadata` in full. `wp_generate_attachment_metadata_custom()` starts from `$metadata = array()` and only sets `width`, `height`, `hwstring_small`, `file`, `sizes` and `image_meta`. Neither `original_image` nor `filesize` appears anywhere in the plugin, so both are dropped every time somebody clicks the button.

`original_image` matters. Since 5.3, an upload bigger than `big_image_size_threshold` (2560px by default) gets scaled, the attachment is pointed at the scaled copy, and the name of the untouched file is stored under that key. It is what `wp_get_original_image_path()` reads, and it is how `wp_delete_attachment()` knows to remove the original. Drop it and WordPress cannot see that file any more. `filesize` has been part of the record since 6.0.

There is a second effect from the same code. Sub sizes are built from `get_attached_file()`, which on a scaled upload is `photo-scaled.jpg`, so they come out as `photo-scaled-300x225.jpg` while the ones the site is serving are `photo-300x225.jpg`. Core builds sub sizes from the original instead, and says why in `wp_create_image_subsizes()`:

```
Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
```

So the rebuild writes a second set of files next to the ones the site is using and repoints the metadata at the new ones. The test image below went from 925,844 bytes on disk to 1,361,848 after a single rebuild.

### Reproduction

WordPress 7.0.2 and PHP 8.4 on the official docker images, no other plugins active, one 4000x3000 jpeg brought in with `wp_insert_attachment()` and `wp_generate_attachment_metadata()`, which is what the uploader does. The rebuild went through `admin-ajax.php` carrying the nonce printed on Tools -> Rebuild Thumbnails, so it is the same request the button sends.

Before, straight after the upload:

```
meta keys      : width,height,file,filesize,sizes,image_meta,original_image
original_image : 'huge.jpg'
filesize       : 112267
wp_get_original_image_path: huge.jpg  (exists)
8 files on disk
```

After one rebuild on 1.14:

```
meta keys      : width,height,hwstring_small,file,sizes,image_meta
original_image : NULL
filesize       : NULL
wp_get_original_image_path: huge-scaled.jpg
14 files on disk (huge-scaled-150x150.jpg and friends, alongside the huge-150x150.jpg set core made)
```

Then `wp_delete_attachment( $id, true )`, which is what Delete Permanently calls:

```
LEFT BEHIND: huge.jpg (378610 bytes), huge-150x150.jpg, huge-300x225.jpg,
             huge-768x576.jpg, huge-1024x768.jpg, huge-1536x1152.jpg, huge-2048x1536.jpg
```

Seven files, including the full resolution original, with no row in the database referring to them.

### The patch

Four changes, all inside `wp_generate_attachment_metadata_custom()`:

- seed `$metadata` from `wp_get_attachment_metadata( $attachment_id, true )` instead of an empty array, so keys the plugin does not manage survive the write. Unfiltered, because it goes straight back into the database.
- set `filesize`, and stop writing `hwstring_small`. That key does not appear anywhere in wp-includes or wp-admin on 7.0.2, and it is absent from image.php, media.php and post.php as far back as 6.2, so nothing reads it.
- resize from `wp_get_original_image_path()` when that file is there, falling back to `$file` when it is not.
- for a size the user did not tick, keep the entry that is already in `$metadata`. The old code re-read it through `image_get_intermediate_size()`, which pulls from that same stored metadata and adds `path` and `url` on top, two keys that core leaves out of the stored record. This branch is unreachable from the UI today anyway, see the last note below.

Both new core functions are behind `function_exists()`, so nothing breaks on older WordPress.

Same site, same image, with the patch applied:

```
meta keys      : width,height,file,filesize,sizes,image_meta,original_image
original_image : 'huge.jpg'
filesize       : 112267
wp_get_original_image_path: huge.jpg  (exists)
8 files before, 8 files after
files rewritten: huge-150x150.jpg, huge-300x225.jpg, huge-768x576.jpg,
                 huge-1024x768.jpg, huge-1536x1152.jpg, huge-2048x1536.jpg
files created  : none
after Delete Permanently: uploads directory is clean
```

### What I checked

Every case below ran twice, once on 1.14 and once on the patch, on the same fresh install.

- 4000x3000 jpeg, all sizes ticked. Metadata now matches what a fresh core upload produces, the six sub sizes are rewritten in place, nothing is orphaned on delete.
- 800x600 jpeg, under the threshold, the ordinary case. Sub sizes rebuilt as before, `filesize` no longer dropped, no extra files, clean delete.
- A PDF. The button is offered on non image attachments, since `attachment_fields_to_edit` runs for them, and on 1.14 the whole record is replaced with an empty array. With the patch it is left alone. My container could not render PDF previews, so there the record was only `filesize`. On a host that does render them the same write takes the `sizes` entries for the previews with it.
- Partial rebuild with only `thumbnail` selected. The 150x150 file is the only one rewritten, and the other five entries plus `original_image` are still in the record afterwards.
- A size the theme no longer registers. Its entry stays in the record, which is what lets WordPress delete the file later. On 1.14 the entry vanished and the file was orphaned.
- A site already damaged by 1.14, then updated. No third set of files appears and the record stays self consistent. The patch does not undo the earlier damage, see below.
- Original missing from disk, `sizes` stored as a string instead of an array, no metadata row at all, and the same rebuild run five times over. No fatals in any of them. The five runs stayed at 8 files and deleted clean. With the original gone the resize falls back to the scaled copy, so that case does still produce the extra set of files, same as 1.14.

It also went through a runtime harness: real WordPress on PHP 8.4, activate, crawl the admin including both of its pages, fire the ajax action and the scheduled events, deactivate and reactivate. 51 requests, PASS, no fatals and no new diagnostics against a stock baseline. 1.14 passes that as well. Nothing here shows up as a crash, which is why the reproduction above is metadata and file listings.

### Things I noticed and left alone

- The size checkboxes do not currently narrow anything. In the handler, `if ( true == $thumbnails )` is true for a non empty array as well, so a partial selection is replaced by the full list of size names and every size gets rebuilt. `! is_array( $thumbnails )` would be the test. Fixing it changes what the button does, so I left it out of a data loss patch.
- Records already damaged by an earlier rebuild are not repaired. `original_image` is gone from them and there is no reliable way to work out which file on disk it was.
- If the original file has been deleted by something else, the fallback resizes the scaled copy, which is the old behaviour and the best available.
- I have not touched the readme or the version number.

### CODE REVIEW

**Comment by RistoNiinemets** [comment]
Thank you for this - all of this has been handled in a complete rewrite of this plugin.

Closing this in favour of the rewrite, with credit to you for the diagnosis — and if you find anything the rewrite got wrong, an issue against 2.0.1 would be very welcome.
